packages/repository-json-schema/src/build-schema.ts

```diff
@@ -91,46 +96,90 @@ export function modelToJsonSchema(ctor: Function): JsonDefinition {
         }
       }
     }
-    if (propCtor && typeof propCtor === 'function') {
-      // errors out if @property.array() is not used on a property of array
-      if (propCtor === Array) {
-        throw new Error('type is defined as an array');
-      }
 
-      const propDef: JsonDefinition = determinePropertyDef(propCtor);
+    // errors out if @property.array() is not used on a property of array
+    if (propCtor === Array) {
+      throw new Error('type is defined as an array');
+    }
 
-      if (!schema.properties) {
-        schema.properties = {};
-      }
+    const propDef: JsonDefinition = determinePropertyDef(propCtor);
 
-      if (propMeta.array === true) {
-        schema.properties[p] = {
-          type: 'array',
-          items: propDef,
-        };
-      } else {
-        schema.properties[p] = propDef;
+    if (!schema.properties) {
+      schema.properties = {};
+    }
+
+    if (propMeta.validationKey) {
+      if (propMeta.validationKey === 'oneOf') {
+        let property: JsonDefinition = (schema.properties[prop] =
+          schema.properties[prop] || {});
+        if (!property.oneOf) {
```

Comment:
AFAIK, juggler does not support `oneOf` constraints. Juggler allows clients to store values violating this constraint and as a result, the REST API will return responses that are not conforming to the OpenAPI spec.

I am proposing to leave `oneOf` out of this pull request. First of all, it's out of MVP scope. Secondly, we should implement support for `oneOf` in juggler first, and only then expose it via decorators.

Please correct me if my understanding is wrong and juggler does support `oneOf` constraint.

Reply:
Interestingly enough ,the acceptance criteria for #786 say

> For @model:
> Add top-level specific metadata to the entry for this type:
> (...)
> oneOf (it's a union type!)

I find this confusing - oneOf is related to `@property`, not `@model` decorator, AFAIK juggler does not support `oneOf` and I don't understand why is this required for our MVP scope? AFAICT, https://github.com/strongloop/loopback-next/tree/master/packages/example-getting-started does not use `oneOf` at all.

@kjdelisle could you please clarify?

Reply:
I'm relatively sure that `oneOf` is not supported juggler as well. Since support of `oneOf` was an MVP goal I've put in support for it and coded under the assumption that not every properties that get put into `@model` decorator had to be juggler related.

I think this just goes to show that we need to extract out the decorators into `@loopback/model` package of some sort to just have a general container for information that juggler and `json-schema` can use. This will require a refactor and I'm not sure when it should be done, but I think it'll save us a lot of headaches in the future.

Reply:
@kjdelisle I see your point about using JSON Schema as the common ground. I have raised this point in the past, but @raymondfeng did not agree IIRC. If we decide to pursue this direction, then we must make it easy to use and tell the user when they are doing something that does not work (like using `oneOf` on a model that's persisted by juggler that does not support `oneOf` validation). 

I also don't see why should `oneOf` be part of our MVP? Remember, we defined the MVP scope as whatever is needed to allow LoopBack4 developers to easily build our [getting-started application](https://github.com/strongloop/loopback-next/tree/master/packages/example-getting-started). AFAICT, this application does not need `oneOf`, therefore `oneOf` is not part of our MVP. 

I consider this as a scope creep, I really wish we could do better in keeping focus on what's absolutely required by MVP. I'm sure it will help us to deliver more of what we committed to in our monthly milestones.

Reply:
> I also don't see why should oneOf be part of our MVP? Remember, we defined the MVP scope as whatever is needed to allow LoopBack4 developers to easily build our getting-started application. AFAICT, this application does not need oneOf, therefore oneOf is not part of our MVP.

I think the task wasn't sufficiently groomed in that it didn't expound upon the expected support for properties and their metadata, so in that respect we definitely misstepped.

All that being said, I'm not certain what we gain by abandoning that code right here and now since it still would be something we'd want to add support for as a part of supporting JSON Schema.

Reply:
>I see your point about using JSON Schema as the common ground. I have raised this point in the past, but @raymondfeng did not agree IIRC. If we decide to pursue this direction, then we must make it easy to use and tell the user when they are doing something that does not work (like using oneOf on a model that's persisted by juggler that does not support oneOf validation).

Okay, I re-read this a few times and I think I get what you're saying now; the fact that the legacy-juggler-bridge doesn't already have any facilities in place for processing this metadata means that it would at best, do nothing and at worst, break things for anyone making use of the LJB, at least until we add logic to ignore/handle it (which is definitely out of scope for this PR, IMO).

Reply:
My problem with this patch is that it creates the impression that people can use `oneOf` restriction when defining their (database) models. When they start using it, they will discover their `oneOf` validation is actually not performed. In the best case, they will do some debugging to find out that `oneOf` is not supported by juggler/database connectors.  My concern is that most people won't do that and they will instead go directly to us, asking what's wrong. This is sort of similar to `basePath` situation which is confusing to LB users.

> All that being said, I'm not certain what we gain by abandoning that code right here and now since it still would be something we'd want to add support for as a part of supporting JSON Schema.

Makes sense. 

I am asking to add a bit of code to our legacy juggler that will throw an exception when somebody tries to add `oneOf` validation to a model property backed by juggler. So that potential users are immediately notified that they are trying to use a feature that's not supported yet.